### PR TITLE
Fix NLB icmp permission duplication

### LINKED ISSUE #56703: AWS NLB service errors with more than one port
**Is this a BUG REPORT or FEATURE REQUEST?**:

 /kind bug

/sig aws

**What happened**:

Create an NLB service on 1.9 with 2 ports, and get this error:

```
I1201 15:16:49.358871       5 aws.go:2753] Adding security group ingress: sg-f8559f84 [{
  FromPort: 30781,
  IpProtocol: "tcp",
  IpRanges: [{
      CidrIp: "0.0.0.0/0",
      Description: "kubernetes.io/rule/nlb/client=a788d91b9d6aa11e7b7f806f096e0fe7"
    }],
  ToPort: 30781
} {
  FromPort: 3,
  IpProtocol: "icmp",
  IpRanges: [{
      CidrIp: "0.0.0.0/0",
      Description: "kubernetes.io/rule/nlb/mtu=a788d91b9d6aa11e7b7f806f096e0fe7"
    }],
  ToPort: 4
} {
  FromPort: 32363,
  IpProtocol: "tcp",
  IpRanges: [{
      CidrIp: "0.0.0.0/0",
      Description: "kubernetes.io/rule/nlb/client=a788d91b9d6aa11e7b7f806f096e0fe7"
    }],
  ToPort: 32363
} {
  FromPort: 3,
  IpProtocol: "icmp",
  IpRanges: [{
      CidrIp: "0.0.0.0/0",
      Description: "kubernetes.io/rule/nlb/mtu=a788d91b9d6aa11e7b7f806f096e0fe7"
    }],
  ToPort: 4
}]
W1201 15:16:49.430537       5 aws.go:2760] Error authorizing security group ingress "InvalidParameterValue: The same permission must not appear multiple times\n\tstatus code: 400, request id: a317f99c-6360-4314-a17c-eeeebfbce0ea"
```

Kubernetes is attempting to create the ICMP rule for MTU discovery once per port, resulting in an error. Creating a service with a single port does not experience this error.

**What you expected to happen**:

The service should be created.

**How to reproduce it (as minimally and precisely as possible)**:

Create a 2 port service in a 1.9 beta cluster.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9 beta1
- Cloud provider or hardware configuration: aws
- OS (e.g. from /etc/os-release): any
- Kernel (e.g. `uname -a`): any

Ping @alebdf @justinsb

### PR DESCRIPTION
**Pull request by aledbf**
**What this PR does / why we need it**:

Fixes an issue with the ICMP rule for MTU during the creation of a NLB

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #56703

### CODE REVIEW

**Comment by aledbf** [comment]
@kubernetes/sig-aws

**Comment by aledbf** [comment]
ping @micahhausler @justinsb

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 1]
```diff
@@ -684,23 +685,28 @@ func (c *Cloud) updateInstanceSecurityGroupsForNLBTraffic(actualGroups []*ec2.Se
...
-						CidrIp:      aws.String(cidr),
-						Description: aws.String(mtuRuleAnnotation),
-					})
-				}
-				permission.IpRanges = ranges
-				if add {
-					adds = append(adds, permission)
-				} else {
-					removes = append(removes, permission)
+				if !icmpPermAdded {
```
Comment (micahhausler) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: Small nit: you could probably move the `mtuRuleAnnotation` into this block

**Review by micahhausler — APPROVED** [review]
LGTM! Thanks for doing this!

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go (hunk 1 above)
Comment (aledbf) [reply pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: done

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 2]
```diff
@@ -684,23 +684,29 @@ func (c *Cloud) updateInstanceSecurityGroupsForNLBTraffic(actualGroups []*ec2.Se
...
+						ranges = append(ranges, &ec2.IpRange{
+							CidrIp:      aws.String(cidr),
+							Description: aws.String(mtuRuleAnnotation),
+						})
+					}
+					permission.IpRanges = ranges
+					if add {
+						adds = append(adds, permission)
+					} else {
+						removes = append(removes, permission)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: If we have 2 ports and we remove one, we don't want to remove the MTU rule, do we?

I suspect the MTU rule belongs outside the port changes loop.

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 3]
```diff
@@ -727,7 +706,26 @@ func (c *Cloud) updateInstanceSecurityGroupsForNLBTraffic(actualGroups []*ec2.Se
...
+		for _, cidr := range clientCidrs {
+			ranges = append(ranges, &ec2.IpRange{
+				CidrIp:      aws.String(cidr),
+				Description: aws.String(mtuRuleAnnotation),
+			})
+		}
+		mtuPermission.IpRanges = ranges
+
 		if len(adds) > 0 {
+			adds = append(adds, mtuPermission)
```
Comment (micahhausler) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: This is not going to work, if a service port was changed from `80` to `8080`, you'd get both an add and remove.

The ICMP rules really need their own tracking structure, you could probably do something similar to the 
`portChanges` map:

```go
// securityGroupID, clientCIDR, add/remove
icmpPortChanges := map[string]map[string]bool{}
```

You'd need to
* get the desired state (which is just an ICMP for each clientCidr for each security group)
* get the actual state (probably just reuse the `for _, actualGroup := range actualGroups` loop)
* append to `adds`/ `removes` as necessary

**Review by micahhausler — CHANGES_REQUESTED** [review]

**Review by micahhausler — APPROVED** [review]
This looks good tome
/lgtm

**Comment by aledbf** [comment]
ping @justinsb

**Comment by justinsb** [comment]
Thanks @aledbf !

**Comment by micahhausler** [comment]
@justinsb How do we get this into the release-1.9 branch?

**Comment by jsafrane** [comment]
> How do we get this into the release-1.9 branch?

Fill release note or tell the bot there is none.

**Comment by rajatjindal** [comment]
Hi @aledbf , do we plan to merge this to 1.9.x? we are running 1.9.6, and running into this issue

**Comment by aledbf** [comment]
@rajatjindal I don't think so. Please check all the comments asking the same thing.

**Pull request merged**